syntheseus/search/algorithms/pdvn.py

```diff
@@ -0,0 +1,372 @@
+
+    Key `node.data` variables used by this algorithm are:
+    - pdvn_mcts_v_syn: accumulated "synthesizability" value (equation 4/7 of Liu et al 2023)
+    - pdvn_mcts_v_cost: accumulated "cost" value (equation 4/7 of Liu et al 2023)
+    - rewards: are tuple valued, with values (syn [binary], cost [float])
+    """
+
+    def __init__(
+        self,
+        c_dead: float,  # cost of a dead end (equation 1 of Liu et al 2023)
+        value_function_syn: BaseNodeEvaluator[OrNode],
+        value_function_cost: BaseNodeEvaluator[OrNode],
+        and_node_cost_fn: BaseNodeEvaluator[AndNode],
+        bound_function: Callable[[AndNode], AndOrGraph] = pucb_bound,  # type: ignore[assignment]  # sloppy typing for bounds
+        **kwargs,
+    ):
+        super().__init__(
+            bound_function=bound_function,  # type: ignore[arg-type]  # sloppy typing for bounds
+            value_function=value_function_cost,
+            reward_function=None,  # type: ignore[arg-type]  # reward function is not used for PDVN MCTS
+            **kwargs,
+        )
+        self.c_dead = c_dead
+        self.value_function_syn = value_function_syn
+        self.value_function_cost = value_function_cost
+        self.and_node_cost_fn = and_node_cost_fn
+        assert self.policy is not None, "PDVN_MCTS requires a policy to be specified."
+
+    def set_node_values(self, nodes, graph):
+        """
+        This function updates values for all nodes in the graph and ensures everything is
+        properly initialized.
+
+        In addition to the logic in the base class (which sets things like depth and has_solution),
+        this function ensures that the synthesizability and cost values are initialized
+        and that the cost of every reaction is set.
+        """
+        updated_nodes = super().set_node_values(nodes, graph)
+
+        # Ensure synthesis and cost values are initialized
+        for node in updated_nodes:
+            node.data.setdefault("pdvn_mcts_v_syn", self.init_mcts_value)
+            node.data.setdefault("pdvn_mcts_v_cost", self.init_mcts_value)
+
+        # Ensure reaction costs are set
+        nodes_without_cost = [
+            n
+            for n in updated_nodes
+            if n.data.get("pdvn_reaction_cost") is None and isinstance(n, AndNode)
+        ]
+        self._set_and_node_costs(nodes_without_cost, graph)
+
+        return updated_nodes
+
+    def _set_and_node_costs(self, and_nodes: Sequence[AndNode], graph: AndOrGraph) -> None:
+        """Helper function to compute the cost of each AndNode using the `and_node_cost_fn`."""
+        costs = self.and_node_cost_fn(and_nodes, graph=graph)
+        assert len(costs) == len(and_nodes)
+        for node, cost in zip(and_nodes, costs):
+            node.data["pdvn_reaction_cost"] = cost
+
+    def choose_successors_to_visit(
+        self,
+        node: ANDOR_NODE,  # type: ignore[override]  # AND/OR vs any graph node
```

Comment:
Fine for now, but @AustinT do you think we should later improve the generics to resolve this? I think this could be done but it might introduce extra complexity (e.g. yet another type variable in the base class)...

Reply:
Yes, the generics for graph types are a big weakness of this package. I really wanted to resolve this before the initial release but couldn't figure out how to do so. If you want to look into it you are very welcome to do so!